pkg/speaker/vip/keepalived.go

```diff
@@ -1,334 +1,344 @@
-				k.log.Error(err, "keepalived configMap delete error")
+			defer logWriter.Close()
+			k.cmd = exec.Command(keepalivedStarter, "--dont-fork", "--log-console", "--log-detail", "--vrrp", "--release-vips")
+			k.cmd.Stdout = logWriter
+			k.cmd.Stderr = logWriter
+			if err := k.cmd.Start(); err != nil {
+				klog.Errorf("Error starting keepalived: %v", err)
+				select {
+				case started <- false:
+				default:
+				}
+				return
+			}
+			select {
+			case started <- true:
+			default:
 			}
+			klog.Infof("Keepalived: started with pid %d", k.cmd.Process.Pid)
+
+			crashCh := make(chan struct{})
+			go func() {
+				if err := k.cmd.Wait(); err != nil {
+					klog.Errorf("Keepalived: crashed, err: %s", err.Error())
+					// Avoid busy loop & hogging CPU resources by waiting before restarting keepalived.
+					time.Sleep(500 * time.Millisecond)
+				}
+				klog.Warning("Keepalived: crashed")
+				close(crashCh)
+			}()
 
-			return
+			<-crashCh
 		}
 	}()
 
-	return nil
-}
+	<-started
 
-// TODO:
-func (l *KeepAlived) ConfigureWithEIP(eip *v1alpha2.Eip, deleted bool) error {
 	return nil
 }
 
-// User can config Keepalived by ConfigMap to specify the images
-// If the ConfigMap exists and the configuration is set, use it,
-//
-//	otherwise, use the default image got from constants.
-func (k *KeepAlived) getImage() string {
-	cm, err := k.client.CoreV1().ConfigMaps(util.EnvNamespace()).
-		Get(context.Background(), constant.OpenELBImagesConfigMap, metav1.GetOptions{})
-	if err != nil {
-		return constant.OpenELBDefaultKeepAliveImage
+func (k *KeepAlived) ConfigureWithEIP(config speaker.Config, deleted bool) error {
+	netif, err := speaker.ParseInterface(config.Iface, true)
+	if err != nil || netif == nil {
+		return err
 	}
 
-	image, exist := cm.Data[constant.OpenELBKeepAliveImage]
-	if !exist {
-		return constant.OpenELBDefaultKeepAliveImage
+	if err := speaker.ValidateInterface(netif, config.IPRange); err != nil {
+		return err
 	}
-	return image
-}
 
-func (k *KeepAlived) getAndCreateConfig() (*corev1.ConfigMap, error) {
-	cm, err := k.client.CoreV1().ConfigMaps(util.EnvNamespace()).Get(context.TODO(), constant.OpenELBVipConfigMap, metav1.GetOptions{})
-	if err != nil {
-		if errors.IsNotFound(err) {
-			return k.createVIPConfigMap()
-		}
-		return nil, err
+	if deleted {
+		delete(k.configs, config.Name)
+	} else {
+		k.configs[config.Name] = &config
 	}
-
-	return cm, nil
+	return nil
 }
 
-func (k *KeepAlived) createVIPConfigMap() (*corev1.ConfigMap, error) {
-	ds, err := k.client.AppsV1().DaemonSets(util.EnvNamespace()).Get(context.TODO(), util.EnvDaemonsetName(), metav1.GetOptions{})
-	if err != nil {
-		k.log.Error(err, fmt.Sprintf("get daemonset:%s error", util.EnvDaemonsetName()))
-		return nil, err
+// Reload sends SIGHUP to keepalived to reload the configuration.
+func (k *KeepAlived) Reload() error {
+	klog.Info("Waiting for keepalived to start")
+	for !k.IsRunning() {
```

Comment:
We should add some info logs here.

Reply:
> We should add some info logs here.

There are already info logs and error logs.